benches/poseidon.rs

```diff
@@ -0,0 +1,255 @@
+            _spec: PhantomData,
+        }
+    }
+
+    fn configure(meta: &mut ConstraintSystem<Fp>) -> Self::Config {
+        let state = (0..WIDTH).map(|_| meta.advice_column()).collect::<Vec<_>>();
+        let partial_sbox = meta.advice_column();
+
+        let rc_a = (0..WIDTH).map(|_| meta.fixed_column()).collect::<Vec<_>>();
+        let rc_b = (0..WIDTH).map(|_| meta.fixed_column()).collect::<Vec<_>>();
+
+        meta.enable_constant(rc_b[0]);
+
+        Self::Config {
+            input: state[..RATE].try_into().unwrap(),
+            poseidon_config: Pow5Chip::configure::<S>(
+                meta,
+                state.try_into().unwrap(),
+                partial_sbox,
+                rc_a.try_into().unwrap(),
+                rc_b.try_into().unwrap(),
+            ),
+        }
+    }
+
+    fn synthesize(
+        &self,
+        config: Self::Config,
+        mut layouter: impl Layouter<Fp>,
+    ) -> Result<(), Error> {
+        let chip = Pow5Chip::construct(config.poseidon_config.clone());
+
+        let message = layouter.assign_region(
+            || "load message",
+            |mut region| {
+                let message_word = |i: usize| {
+                    let value = self.message.map(|message_vals| message_vals[i]);
+                    let cell = region.assign_advice(
+                        || format!("load message_{}", i),
+                        config.input[i],
+                        0,
+                        || value.ok_or(Error::Synthesis),
+                    )?;
+                    Ok(CellValue::new(cell, value))
+                };
+
+                let message: Result<Vec<_>, Error> = (0..RATE).map(message_word).collect();
+                Ok(message?.try_into().unwrap())
+            },
+        )?;
+
+        let hasher = Hash::<_, _, S, _, WIDTH, RATE>::init(
+            chip,
+            layouter.namespace(|| "init"),
+            ConstantLength::<RATE>,
+        )?;
+        let output = hasher.hash(layouter.namespace(|| "hash"), message)?;
+
+        layouter.assign_region(
+            || "constrain output",
+            |mut region| {
+                let expected_var = region.assign_advice(
+                    || "load output",
+                    config.input[0],
+                    0,
+                    || self.output.ok_or(Error::Synthesis),
+                )?;
+                region.constrain_equal(output.cell(), expected_var)
+            },
+        )
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+struct MySpec<const WIDTH: usize, const RATE: usize>;
+
+impl Spec<Fp, 3, 2> for MySpec<3, 2> {
+    fn full_rounds() -> usize {
+        8
+    }
+
+    fn partial_rounds() -> usize {
+        56
+    }
+
+    fn sbox(val: Fp) -> Fp {
+        val.pow_vartime(&[5])
+    }
+
+    fn secure_mds() -> usize {
+        0
+    }
+}
+
+impl Spec<Fp, 9, 8> for MySpec<9, 8> {
+    fn full_rounds() -> usize {
+        8
+    }
+
+    fn partial_rounds() -> usize {
+        56
+    }
```

Comment:
These aren't necessarily correct; IIRC we are supposed to derive the number of full and partial rounds from the desired permutation width etc. using the reference Sage code. We should separately (maybe in the refactor PR) document how to do this on the `Spec` trait itself.